File: js/2_1_proxy.js
class RealSubject {
    request() {
        return "Real subject response";
    }
}

class MyProxy {
    constructor(realSubject) {
        if (realSubject && realSubject.request)
            this.realSubject = realSubject;
        else
            this.realSubject = new RealSubject();
    }

    request() {
        if (this.checkAccess()) {
            const response = this.realSubject.request();
            this.logAccess(response);
            return response;
        }
        else
            return "Proxy response";
    }

    checkAccess() {
        return Math.random() < 0.5;
    }

    logAccess(message) {
        console.log(`Request was proxied: ${message}`);
    }
}

function ProxyFunction(realSubject) {
    function proxyMethod(method) {
        return function () {
            if (checkAccess()) {
                const response = method.apply(realSubject);
                logAccess(response);
                return response;
            }
            else
                return "Proxy response";
        }

        function checkAccess() {
            return  Math.random() < 0.5;
        }

        function logAccess(message) {
            console.log(`Request was proxied by function: ${message}`);
        }
    }

    realSubject.request = proxyMethod(realSubject.request);
    return realSubject;
}

export { RealSubject, MyProxy, ProxyFunction }

File: cs/2_1_proxy.cs
using System;
namespace Structural.Proxy
{
    public interface ISubject
    {
        string Request();
    }

    class RealSubject : ISubject
    {
        private string Ip;

        public RealSubject(string Ip)
        {
            this.Ip = Ip;
        }
        public string Request()
        {
            return "Real subject response";
        }
    }

    class Proxy : ISubject
    {
        private RealSubject _realSubject;
        public Proxy(RealSubject realSubject)
        {
            this._realSubject = realSubject;
        }
        public Proxy(string Ip)
        {
            this._realSubject = new RealSubject(Ip);
        }

        public string Request()
        {
            if (this.CheckAccess())
            {
                string response = this._realSubject.Request();
                this.LogAccess(response);
                return response;
            }
            else
                return "Proxy response";
        }

        private Random rnd = new Random(2);
        private bool CheckAccess()
        {
            return rnd.NextDouble() < 0.5;
        }

        private void LogAccess(string message)
        {
            Console.WriteLine($"Request was proxied: {message}");
        }
    }
}